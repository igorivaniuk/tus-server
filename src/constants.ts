export const REQUEST_METHODS = ['POST', 'HEAD', 'PATCH', 'OPTIONS']

export const HEADERS = [
  'Authorization',
  'Content-Type',
  'Location',
  'Tus-Extension',
  'Tus-Max-Size',
  'Tus-Resumable',
  'Tus-Version',
  'Upload-Defer-Length',
  'Upload-Length',
  'Upload-Metadata',
  'Upload-Offset',
  'X-HTTP-Method-Override',
  'X-Requested-With',
]

export const HEADERS_LOWERCASE = HEADERS.map(header => header.toLowerCase())

export const ALLOWED_HEADERS = HEADERS.join(', ')
export const ALLOWED_METHODS = REQUEST_METHODS.join(', ')
export const EXPOSED_HEADERS = HEADERS.join(', ')
export const MAX_AGE = 86400
export const TUS_RESUMABLE = '1.0.0'
export const TUS_VERSION = ['1.0.0']
