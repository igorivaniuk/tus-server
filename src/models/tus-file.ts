export class TusFile {
  public id = this.fileId

  public createdAt = new Date()

  constructor(
    public fileId: string,
    public uploadLength: number,
    public uploadDeferLength: number,
    public uploadMetadata: string,
  ) {
    if (!fileId) {
      throw new Error('[TusFile] constructor must be given a fileId')
    }

    if (uploadLength === undefined && uploadDeferLength === undefined) {
      throw new Error('[TusFile] constructor must be given either a uploadLength or uploadDeferLength')
    }
  }

  /**
   *
   * @param period in seconds
   */
  getExpiredAt(period: number): Date {
    return new Date(new Date(this.createdAt).getTime() + period * 1000)
  }
}
