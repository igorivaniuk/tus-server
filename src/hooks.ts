import { TusFile } from './models/tus-file'
import * as debug from 'debug'
const log = debug('tus-server:hooks')

export enum HookType {
  BeforeCreate = 'beforeCreate',
  AfterCreate = 'afterCreate',
  AfterComplete = 'afterComplete',
}

export class Hooks<T> {
  private _hooks = {
    [HookType.BeforeCreate]: [] as any[],
    [HookType.AfterCreate]: [] as any[],
    [HookType.AfterComplete]: [] as any[],
  }

  beforeCreate(fn: (req: T) => Promise<any>) {
    this._hooks[HookType.BeforeCreate].push(fn)
  }

  afterCreate(fn: (file: TusFile, req: T) => Promise<any>) {
    this._hooks[HookType.AfterCreate].push(fn)
  }

  afterComplete(fn: (file: TusFile, req: T) => Promise<any>) {
    this._hooks[HookType.AfterComplete].push(fn)
  }

  async run(type: HookType, ...args: any[]) {
    log('run', type)
    for(let fn of this._hooks[type]) {
      await fn(...args)
    }
  }
}
