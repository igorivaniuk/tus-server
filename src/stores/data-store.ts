import debug from 'debug'
import * as http from 'http'
import * as uuid from 'uuid'
import { InvalidLengthException } from '../errors/invalid-length.exception'
import { Hooks, HookType } from '../hooks'
import { TusFile } from '../models/tus-file'

const log = debug('tus-server:stores')

export interface BytesWritten {
  size: number
  upload_length: number
  upload_defer_length?: number
  upload_metadata?: string
}

export interface DataStoreOptions {
  namingFunction?: (req: http.IncomingMessage) => string
}

export abstract class DataStore<T extends http.IncomingMessage = http.IncomingMessage> {
  _extensions: string[] = []

  options: Required<DataStoreOptions>

  expiration = 0

  readonly hooks = new Hooks<T>()

  constructor(options: DataStoreOptions) {
    this.options = {
      ...options,
      namingFunction: () => uuid.v4(),
    }
  }

  get extensions() {
    if (this._extensions.length === 0) {
      return null
    }
    return this._extensions.join()
  }

  async beforeCreate(req: T) {
    await this.hooks.run(HookType.BeforeCreate, req)
  }

  async afterCreate(file: TusFile, req: T) {
    await this.hooks.run(HookType.AfterCreate, file, req)
  }

  async afterComplete(file: TusFile, req: T) {
    await this.hooks.run(HookType.AfterComplete, file, req)
  }

  async createFile(req: T) {
    log('[DataStore] create')
    const uploadLength = req.headers['upload-length']
    // const uploadDeferLength = req.headers['upload-defer-length']
    const uploadMetadata = req.headers['upload-metadata'] as string

    if (uploadLength === undefined) {
      throw new InvalidLengthException()
    }

    const fileId = this.options.namingFunction(req)
    return new TusFile(fileId, +uploadLength!, 0, uploadMetadata)
  }

  abstract async init?(): Promise<void>;

  /**
   * Called in POST requests. This method just creates a
   * file, implementing the creation extension.
   *
   * http://tus.io/protocols/resumable-upload.html#creation
   *
   * @param  {object} req http.incomingMessage
   * @return {Promise}
   */
  abstract async create(req: T): Promise<TusFile>
  /**
   * Called in PATCH requests. This method should write data
   * to the DataStore file, and possibly implement the
   * concatenation extension.
   *
   * http://tus.io/protocols/resumable-upload.html#concatenation
   *
   */
  abstract async write(req: T, fileId: string, offset: number): Promise<number>

  /**
   * Called in HEAD requests. This method should return the bytes
   * writen to the DataStore, for the client to know where to resume
   * the upload.
   *
   */
  abstract async getOffset(fileId: string): Promise<BytesWritten>
}
