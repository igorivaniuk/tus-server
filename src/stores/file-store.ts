import * as debug from 'debug'
import * as fs from 'fs'
import * as http from 'http'
import { ConfigStore } from '../config-stores/config-store.interface'
import { MemoryConfigStore } from '../config-stores/memory-config-store'
import { FileNoLongerExistsException } from '../errors/file-no-longer-exists.exception'
import { FileWriteException } from '../errors/file-write.exception'
import { NotFoundException } from '../errors/not-found.exception'
import { TusException } from '../errors/tus.exception'
import { TusFile } from '../models/tus-file'
import { BytesWritten, DataStore, DataStoreOptions } from './data-store'

const log = debug('tus-server:file-store')
const MASK = '0777'
const IGNORED_MKDIR_ERROR = 'EEXIST'
const FILE_DOESNT_EXIST = 'ENOENT'

export interface FileStoreOptions extends DataStoreOptions {
  directory?: string
  path?: string
  configStore?: ConfigStore
  // time in seconds
  expiration?: number
}

export class FileStore<T extends http.IncomingMessage = http.IncomingMessage> extends DataStore<T> {
  _extensions = ['creation']
  directory: string
  configStore: ConfigStore

  // default 0 - is disabled
  expiration = 0

  constructor(options: FileStoreOptions) {
    super(options)
    if (!options.path && !options.directory) {
      throw new TusException('Bed file store config')
    }

    this.directory = options.directory || options.path!.replace(/^\//, '')
    this.configStore = options.configStore || new MemoryConfigStore()

    if (options.expiration) {
      this.expiration = options.expiration
      this._extensions.push('expiration')
    }
  }

  async init(): Promise<void> {
    await this._checkOrCreateDirectory()
  }

  /**
   *  Ensure the directory exists.
   */
  _checkOrCreateDirectory() {
    return new Promise((resolve, reject) => {
      fs.mkdir(this.directory, MASK, error => {
        if (error && error.code !== IGNORED_MKDIR_ERROR) {
          return reject(error)
        }
        resolve()
      })
    })
  }

  filePath(fileId: string) {
    return `${this.directory}/${fileId}`
  }

  /**
   * Create an empty file.
   *
   */
  async create(req: T) {
    log('create')
    await this.beforeCreate(req)

    const file = await this.createFile(req)
    await this.touch(file)
    await this.configStore.set(file.id, file)

    await this.afterCreate(file, req)
    return file
  }

  async touch(file: TusFile) {
    let fd = await fs.promises.open(this.filePath(file.id), 'w')
    await fd.close()
  }

  /**
   * Write to the file, starting at the provided offset
   *
   * @param  {object} req http.IncomingMessage
   * @param  {string} fileId   Name of file
   * @param  {number} offset   starting offset
   * @return {Promise}
   */
  async write(req: T, fileId: string, offset: number) {
    const path = this.filePath(fileId)
    const config = await this.configStore.get(fileId)
    const options = {
      flags: 'r+',
      start: offset,
    }
    let newOffset = 0

    const stream = fs.createWriteStream(path, options)
    req.on('data', buffer => {
      newOffset += buffer.length
    })

    return new Promise<number>((resolve, reject) => {
      stream.on('error', e => {
        log('[FileStore] write: Error', e)
        reject(new FileWriteException())
      })

      req.pipe(stream).on('finish', () => {
        log(`[FileStore] write: ${newOffset} bytes written to ${path}`)
        offset += newOffset
        log(`[FileStore] write: File is now ${offset} bytes`)

        if (config && config.uploadLength === offset) {
          return this.afterComplete(config, req)
            .then(() => resolve(offset))
            .catch(reject)
        }
        resolve(offset)
      })
    })
  }

  /**
   * Return file stats, if they exits
   *
   * @param  {string} fileId name of the file
   * @return {object}           fs stats
   */
  async getOffset(fileId: string): Promise<BytesWritten> {
    const filePath = this.filePath(fileId)
    const config = await this.configStore.get(fileId)
    if (!config) {
      throw new NotFoundException()
    }

    return new Promise<BytesWritten>((resolve, reject) => {
      fs.stat(filePath, (error, stats) => {
        if (error && error.code === FILE_DOESNT_EXIST && config) {
          log(`[FileStore] getOffset: No file found at ${filePath} but db record exists`, config)
          return reject(new FileNoLongerExistsException())
        }

        if (error) {
          return reject(error)
        }

        if (stats.isDirectory()) {
          log(`[FileStore] getOffset: ${filePath} is a directory`)
          return reject(new NotFoundException())
        }

        return resolve({
          size: stats.size,
          upload_length: config.uploadLength,
          upload_defer_length: config.uploadDeferLength,
          upload_metadata: config.uploadMetadata,
        })
      })
    })
  }
}
