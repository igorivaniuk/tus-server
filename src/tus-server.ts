import debug from 'debug'
import * as http from 'http'
import { EXPOSED_HEADERS, TUS_RESUMABLE } from './constants'
import { TusException } from './errors/tus.exception'
import { BaseHandler, HandleResult } from './handlers/base.handler'
import { HeadHandler } from './handlers/head.handler'
import { OptionsHandler } from './handlers/options.handler'
import { PatchHandler } from './handlers/patch.handler'
import { PostHandler } from './handlers/post.handler'
import { Hooks } from './hooks'
import { DataStore } from './stores/data-store'
import { RequestValidator } from './validators/request-validator'

const log = debug('tus-server:server')

export interface TusServerConfig {
  dataStore: DataStore
  maxSize?: number
}

export class TusServer<T extends http.IncomingMessage = http.IncomingMessage> {
  readonly handlers: { [method: string]: BaseHandler }

  constructor(private config: TusServerConfig) {
    this.handlers = {
      HEAD: new HeadHandler(config),
      OPTIONS: new OptionsHandler(config),
      PATCH: new PatchHandler(config),
      POST: new PostHandler(config),
    }

    if (typeof config.dataStore.init === 'function') {
      config.dataStore.init().catch(err => console.error('Init store error:', err))
    }
  }

  get hooks() {
    return this.config.dataStore.hooks as Hooks<T>
  }

  /**
   * Main server requestListener, invoked on every 'request' event.
   *
   * @param  {object} req http.IncomingMessage
   * @param  {object} res http.ServerResponse
   */
  handle<T extends http.IncomingMessage = http.IncomingMessage>(
    req: T,
    res: http.ServerResponse,
  ): Promise<void> | void {
    log(`handle: ${req.method} ${req.url}`)

    // Allow overriding the HTTP method. The reason for this is
    // that some libraries/environments to not support PATCH and
    // DELETE requests, e.g. Flash in a browser and parts of Java
    let overrideMethod = req.headers['x-http-method-override']
    if (typeof overrideMethod === 'string') {
      req.method = overrideMethod.toUpperCase()
    }

    // The Tus-Resumable header MUST be included in every request and
    // response except for OPTIONS requests. The value MUST be the version
    // of the protocol used by the Client or the Server.
    res.setHeader('Tus-Resumable', TUS_RESUMABLE)
    if (req.method !== 'OPTIONS' && req.headers['tus-resumable'] === undefined) {
      res.writeHead(412, 'Precondition Failed')
      return res.end('Tus-Resumable Required\n')
    }

    // Validate all required headers to adhere to the tus protocol
    const invalidHeaders = []
    for (const headerName in req.headers) {
      if (req.method === 'OPTIONS') {
        break
      }

      if (RequestValidator.isInvalidHeader(headerName, req.headers[headerName] as string)) {
        log(`Invalid ${headerName} header: ${req.headers[headerName]}`)
        invalidHeaders.push(headerName)
      }
    }

    if (invalidHeaders.length > 0) {
      // The request was not configured to the tus protocol
      res.writeHead(412, 'Precondition Failed')
      return res.end(`Invalid ${invalidHeaders.join(' ')}\n`)
    }

    // Enable CORS
    res.setHeader('Access-Control-Expose-Headers', EXPOSED_HEADERS)
    if (req.headers.origin) {
      res.setHeader('Access-Control-Allow-Origin', req.headers.origin)
    }

    // Invoke the handler for the method requested
    let handler = this.handlers[req.method!] as BaseHandler
    if (handler) {
      return handler
        .handle(req)
        .then(result => this.writeResult(res, result))
        .catch(err => this.writeError(res, err))
    }

    // 404 Anything else
    res.writeHead(404, {})
    res.write('Not found\n')
    return res.end()
  }

  writeError(res: http.ServerResponse, error: any) {
    log(`writeError`)
    if (error instanceof TusException) {
      let response = error.getResponse()
      if (typeof response === 'string') {
        res.writeHead(error.getStatus(), {
          'Content-Type': 'text/plain',
        })
        return res.end(response)
      }
      try {
        let body = JSON.stringify(response)
        res.writeHead(error.getStatus(), {
          'Content-Type': 'application/json',
        })
        res.end(body)
      } catch (e) {
        log('stringify error', e)
        res.writeHead(500, {
          'Content-Type': 'text/plain',
        })
        res.end('Server error')
      }
      return
    }

    log('[ERROR]', error)

    res.writeHead(500)
    res.end()
  }

  writeResult(res: http.ServerResponse, result: HandleResult) {
    log(`writeResult`)
    const headers: any = {}
    for (const key of Array.from(result.headers.keys())) {
      headers[key] = result.headers.get(key)
    }
    if (result.body) {
      headers['Content-Length'] = result.body.length
    }

    res.writeHead(result.statusCode, headers)
    if (result.body) {
      res.write(result.body)
    }
    return res.end()
  }

  listen() {
    const server = http.createServer((req, res) => this.handle(req, res))
    return server.listen.apply(server, arguments as any)
  }
}
