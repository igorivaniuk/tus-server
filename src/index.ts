import { TusServer } from './tus-server'
import { FileStore } from './stores/file-store'
import { TusFile } from './models/tus-file'

export { TusServer, FileStore, TusFile }
