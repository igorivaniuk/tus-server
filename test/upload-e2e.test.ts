// process.env.DEBUG = 'tus*'
import { exec } from 'child_process'
import * as fs from 'fs'
import * as http from 'http'
import { Server } from 'http'
import * as path from 'path'
import * as request from 'supertest'
import { TUS_RESUMABLE } from '../src/constants'
import { TusException } from '../src/errors/tus.exception'
import { TusFile } from '../src/models/tus-file'
import { FileStore } from '../src/stores/file-store'
import { TusServer } from '../src/tus-server'

const TEST_FILE_SIZE = 960244
const MAX_SIZE = TEST_FILE_SIZE + 100
const TEST_FILE_PATH = path.resolve(__dirname, 'test.mp4')
const TEST_METADATA = 'some data, for you'

describe('Upload', () => {
  let agent: request.SuperTest<request.Test>
  let tus: TusServer
  let app: Server
  let fileStore: FileStore
  beforeEach(async () => {
    fileStore = new FileStore({
      directory: '/tmp/tus',
      expiration: 24 * 3600,
    })
    tus = new TusServer({
      dataStore: fileStore,
      maxSize: MAX_SIZE,
    })
    app = tus.listen()
    agent = request.agent(app)
  })

  afterEach(done => {
    exec(`rm -r /tmp/tus`, err => {
      if (err) {
        return done(err)
      }

      return done()
    })
    app.close()
  })

  describe('OPTIONS', () => {
    it('should return server config', async () => {
      let res: any = await agent.options(`/`).set('Tus-Resumable', TUS_RESUMABLE)
      expect(res.headers).toMatchSnapshot({
        date: expect.any(String),
      })
    })
  })

  describe('HEAD', () => {
    let file: TusFile
    beforeEach(async () => {
      file = await fileStore.create({
        headers: {
          'upload-length': 100,
          'upload-metadata': 'test test',
        },
      } as any)
    })

    it('should return 404 for', async () => {
      let res = await agent.head('/blablabla').set('Tus-Resumable', TUS_RESUMABLE)
      expect(res.status).toBe(404)
    })

    it('tus version', async () => {
      let res = await agent.head('/blablabla').set('Tus-Resumable', TUS_RESUMABLE).set('Tus-Version', '0.2.2')
      expect(res.status).toBe(412)
    })

    it('the Upload-Defer-Length value MUST be 1', async () => {
      let res = await agent.head('/blablabla').set('Tus-Resumable', TUS_RESUMABLE).set('Upload-Defer-Length', '1234')
      expect(res.status).toBe(412)
    })

    it('should return a starting offset, metadata for the new file', async () => {
      let res = await agent.head(`/uploads/${file.id}`).set('Tus-Resumable', TUS_RESUMABLE)
      expect(res.status).toBe(200)
      expect(res.get('tus-resumable')).toBe(TUS_RESUMABLE)
      expect(res.get('upload-offset')).toBe('0')
      expect(res.get('upload-length')).toBe('100')
      expect(res.get('upload-metadata')).toBe('test test')
    })

    it('should return 410 Gone for the file that has been deleted', async () => {
      fs.unlinkSync(fileStore.filePath(file.id))
      let res = await agent.head(`/uploads/${file.id}`).set('Tus-Resumable', TUS_RESUMABLE)
      expect(res.status).toBe(410)
      expect(res.header['tus-resumable']).toBe(TUS_RESUMABLE)
    })

    it('should return 404 for the file that has been deleted and no config', async () => {
      fs.unlinkSync(fileStore.filePath(file.id))
      await fileStore.configStore.clear()
      let res = await agent.head(`/uploads/${file.id}`).set('Tus-Resumable', TUS_RESUMABLE)
      expect(res.status).toBe(404)
      expect(res.header['tus-resumable']).toBe(TUS_RESUMABLE)
    })
  })

  describe('POST', () => {
    it('should response with 201', async () => {
      let res = await agent
        .post('/uploads/')
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Length', String(TEST_FILE_SIZE))
        .set('Upload-Metadata', TEST_METADATA)

      expect(res.status).toBe(201)
      expect(res.header['location']).toMatch(
        /^\/uploads\/[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89aAbB][a-f0-9]{3}-[a-f0-9]{12}$/,
      )
      expect(res.header['tus-resumable']).toBe(TUS_RESUMABLE)
    })

    it('should response with 413 if max file size', async () => {
      let res = await agent
        .post('/uploads/')
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Length', String(MAX_SIZE + 100))
        .set('Upload-Metadata', TEST_METADATA)
      expect(res.status).toBe(413)
      expect(res.text).toBe('Request Entity Too Large')
    })

    it('should response with 412 empty upload-length', async () => {
      let res = await agent.post('/uploads/').set('Tus-Resumable', TUS_RESUMABLE)
      expect(res.status).toBe(412)
      expect(res.text).toBe('Upload-Length or Upload-Defer-Length header required')
    })

    it('should response with hook error', async () => {
      tus.hooks.beforeCreate(async (req: any) => {
        throw new TusException('Bad request', 400)
      })
      let res = await agent
        .post('/uploads/')
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Length', String(TEST_FILE_SIZE))
        .set('Upload-Metadata', TEST_METADATA)
      expect(res.status).toBe(400)
      expect(res.text).toBe('Bad request')
    })

    it('should response with hook error 1', async () => {
      tus.hooks.afterCreate(async (req: any) => {
        throw new TusException('Bad request', 401)
      })
      let res = await agent
        .post('/uploads/')
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Length', String(TEST_FILE_SIZE))
        .set('Upload-Metadata', TEST_METADATA)
      expect(res.status).toBe(401)
      expect(res.text).toBe('Bad request')
    })
  })

  describe('PATCH', () => {
    it('should 404 paths without a file id', done => {
      agent
        .patch(`/uploads/`)
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Offset', '0')
        .set('Content-Type', 'application/offset+octet-stream')
        .expect(404)
        .expect('Tus-Resumable', TUS_RESUMABLE)
        .end(done)
    })

    it('should 403 missing header Upload-Offset', done => {
      agent
        .patch(`/uploads/dont_exist`)
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Content-Type', 'application/offset+octet-stream')
        .expect(403)
        .expect('Tus-Resumable', TUS_RESUMABLE)
        .end(done)
    })

    it('should 403 missing header Content-Type', done => {
      agent
        .patch(`/uploads/dont_exist`)
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Offset', '0')
        .expect(403)
        .expect('Tus-Resumable', TUS_RESUMABLE)
        .end(done)
    })

    it('should 404 paths that do not exist', done => {
      agent
        .patch(`/uploads/dont_exist`)
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Offset', '0')
        .set('Content-Type', 'application/offset+octet-stream')
        .expect(404)
        .expect('Tus-Resumable', TUS_RESUMABLE)
        .end(done)
    })

    it('should upload the file', async () => {
      let spy = jest.fn()
      tus.hooks.afterComplete(spy)
      let res = await agent
        .post('/uploads')
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Length', String(TEST_FILE_SIZE))
        .set('Upload-Metadata', TEST_METADATA)

      expect(res.status).toBe(201)
      let fileId = res.header['location'].split('/').pop()

      const read_stream = fs.readFileSync(TEST_FILE_PATH)
      const write_stream = agent
        .patch(`/uploads/${fileId}`)
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Offset', '0')
        .set('Content-Type', 'application/offset+octet-stream')

      let res2 = await new Promise<any>((resolve, reject) => {
        write_stream.on('response', res => {
          resolve(res)
        })
        write_stream.on('error', reject)

        write_stream.write(read_stream)
        write_stream.end(() => {})
      })

      expect(res2.status).toBe(204)
      expect(res2.headers['tus-resumable']).toBe(TUS_RESUMABLE)
      expect(res2.headers['upload-offset']).toBe(String(TEST_FILE_SIZE))
      expect(spy).toBeCalled()
      expect(spy).toBeCalledTimes(1)
    })
  })
})
