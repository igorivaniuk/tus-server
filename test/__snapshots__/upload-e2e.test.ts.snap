// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload OPTIONS should return server config 1`] = `
Object {
  "access-control-allow-headers": "Authorization, Content-Type, Location, Tus-Extension, Tus-Max-Size, Tus-Resumable, Tus-Version, Upload-Defer-Length, Upload-Length, Upload-Metadata, Upload-Offset, X-HTTP-Method-Override, X-Requested-With",
  "access-control-allow-methods": "POST, HEAD, PATCH, OPTIONS",
  "access-control-expose-headers": "Authorization, Content-Type, Location, Tus-Extension, Tus-Max-Size, Tus-Resumable, Tus-Version, Upload-Defer-Length, Upload-Length, Upload-Metadata, Upload-Offset, X-HTTP-Method-Override, X-Requested-With",
  "access-control-max-age": "86400",
  "connection": "close",
  "date": Any<String>,
  "tus-extension": "creation,expiration",
  "tus-max-size": "960344",
  "tus-resumable": "1.0.0",
  "tus-version": "1.0.0",
}
`;
